File: src/motionprint/geometry.py
"""Procedural 3D mesh generation and deformation.

Generates icosphere, torus, and superellipsoid meshes, plus hash-driven
vertex deformation for creating unique organic shapes.
"""

from __future__ import annotations

import dataclasses
import math

import numpy as np


@dataclasses.dataclass
class MeshData:
    """Triangle mesh with per-vertex positions and normals."""

    vertices: np.ndarray   # (N, 3) float32
    normals: np.ndarray    # (N, 3) float32
    indices: np.ndarray    # (M, 3) uint32

    def interleaved(self) -> np.ndarray:
        """Return (N, 6) float32 array: [x,y,z, nx,ny,nz] per vertex."""
        return np.hstack([self.vertices, self.normals]).astype(np.float32)

    def flat_indices(self) -> np.ndarray:
        """Return (M*3,) uint32 array for element buffer."""
        return self.indices.flatten().astype(np.uint32)


def _compute_normals(vertices: np.ndarray, indices: np.ndarray) -> np.ndarray:
    """Compute per-vertex normals by averaging adjacent face normals."""
    normals = np.zeros_like(vertices, dtype=np.float64)
    v0 = vertices[indices[:, 0]]
    v1 = vertices[indices[:, 1]]
    v2 = vertices[indices[:, 2]]
    face_normals = np.cross(v1 - v0, v2 - v0)
    for i in range(3):
        np.add.at(normals, indices[:, i], face_normals)
    lengths = np.linalg.norm(normals, axis=1, keepdims=True)
    lengths = np.maximum(lengths, 1e-10)
    return (normals / lengths).astype(np.float32)


def icosphere(subdivision: int = 3) -> MeshData:
    """Generate an icosphere by subdividing an icosahedron.

    Args:
        subdivision: Number of subdivision levels (0=icosahedron, 3=642 verts).
    """
    # Golden ratio
    t = (1.0 + math.sqrt(5.0)) / 2.0

    # 12 vertices of a regular icosahedron
    verts = [
        (-1, t, 0), (1, t, 0), (-1, -t, 0), (1, -t, 0),
        (0, -1, t), (0, 1, t), (0, -1, -t), (0, 1, -t),
        (t, 0, -1), (t, 0, 1), (-t, 0, -1), (-t, 0, 1),
    ]

    # 20 triangular faces
    faces = [
        (0, 11, 5), (0, 5, 1), (0, 1, 7), (0, 7, 10), (0, 10, 11),
        (1, 5, 9), (5, 11, 4), (11, 10, 2), (10, 7, 6), (7, 1, 8),
        (3, 9, 4), (3, 4, 2), (3, 2, 6), (3, 6, 8), (3, 8, 9),
        (4, 9, 5), (2, 4, 11), (6, 2, 10), (8, 6, 7), (9, 8, 1),
    ]

    # Normalize initial vertices to unit sphere
    vertices = []
    for v in verts:
        length = math.sqrt(v[0] ** 2 + v[1] ** 2 + v[2] ** 2)
        vertices.append((v[0] / length, v[1] / length, v[2] / length))

    # Subdivide
    midpoint_cache: dict[tuple[int, int], int] = {}

    def get_midpoint(i1: int, i2: int) -> int:
        key = (min(i1, i2), max(i1, i2))
        if key in midpoint_cache:
            return midpoint_cache[key]
        v1, v2 = vertices[i1], vertices[i2]
        mid = ((v1[0] + v2[0]) / 2, (v1[1] + v2[1]) / 2, (v1[2] + v2[2]) / 2)
        length = math.sqrt(mid[0] ** 2 + mid[1] ** 2 + mid[2] ** 2)
        mid = (mid[0] / length, mid[1] / length, mid[2] / length)
        idx = len(vertices)
        vertices.append(mid)
        midpoint_cache[key] = idx
        return idx

    triangles = list(faces)
    for _ in range(subdivision):
        midpoint_cache.clear()
        new_triangles = []
        for tri in triangles:
            a = get_midpoint(tri[0], tri[1])
            b = get_midpoint(tri[1], tri[2])
            c = get_midpoint(tri[2], tri[0])
            new_triangles.extend([
                (tri[0], a, c),
                (tri[1], b, a),
                (tri[2], c, b),
                (a, b, c),
            ])
        triangles = new_triangles

    verts_arr = np.array(vertices, dtype=np.float32)
    idx_arr = np.array(triangles, dtype=np.uint32)
    normals = verts_arr.copy()  # For a unit sphere, normals equal positions

    return MeshData(vertices=verts_arr, normals=normals, indices=idx_arr)


def torus(major_radius: float = 1.0, minor_radius: float = 0.4,
          major_segments: int = 48, minor_segments: int = 24) -> MeshData:
    """Generate a torus mesh."""
    vertices = []
    indices = []

    for i in range(major_segments):
        theta = 2.0 * math.pi * i / major_segments
        cos_t, sin_t = math.cos(theta), math.sin(theta)
        for j in range(minor_segments):
            phi = 2.0 * math.pi * j / minor_segments
            cos_p, sin_p = math.cos(phi), math.sin(phi)
            r = major_radius + minor_radius * cos_p
            x = r * cos_t
            y = r * sin_t
            z = minor_radius * sin_p
            vertices.append((x, y, z))

    for i in range(major_segments):
        for j in range(minor_segments):
            curr = i * minor_segments + j
            next_j = i * minor_segments + (j + 1) % minor_segments
            next_i = ((i + 1) % major_segments) * minor_segments + j
            next_ij = ((i + 1) % major_segments) * minor_segments + (j + 1) % minor_segments
            indices.append((curr, next_i, next_ij))
            indices.append((curr, next_ij, next_j))

    verts_arr = np.array(vertices, dtype=np.float32)
    idx_arr = np.array(indices, dtype=np.uint32)
    normals = _compute_normals(verts_arr, idx_arr)

    return MeshData(vertices=verts_arr, normals=normals, indices=idx_arr)


def superellipsoid(e1: float = 1.0, e2: float = 1.0,
                   u_segments: int = 32, v_segments: int = 32) -> MeshData:
    """Generate a superellipsoid mesh.

    e1, e2 control the shape: (1,1)=sphere, (0.1,0.1)≈cube, (2,2)=octahedron-ish.
    """
    def _sign_pow(base: float, exp: float) -> float:
        if base == 0:
            return 0.0
        sign = 1.0 if base > 0 else -1.0
        return sign * abs(base) ** exp

    vertices = []
    for i in range(u_segments + 1):
        u = -math.pi / 2 + math.pi * i / u_segments
        cos_u, sin_u = math.cos(u), math.sin(u)
        for j in range(v_segments + 1):
            v = -math.pi + 2 * math.pi * j / v_segments
            cos_v, sin_v = math.cos(v), math.sin(v)
            x = _sign_pow(cos_u, e1) * _sign_pow(cos_v, e2)
            y = _sign_pow(cos_u, e1) * _sign_pow(sin_v, e2)
            z = _sign_pow(sin_u, e1)
            vertices.append((x, y, z))

    indices = []
    for i in range(u_segments):
        for j in range(v_segments):
            row = v_segments + 1
            curr = i * row + j
            indices.append((curr, curr + row, curr + row + 1))
            indices.append((curr, curr + row + 1, curr + 1))

    verts_arr = np.array(vertices, dtype=np.float32)
    idx_arr = np.array(indices, dtype=np.uint32)
    normals = _compute_normals(verts_arr, idx_arr)

    return MeshData(vertices=verts_arr, normals=normals, indices=idx_arr)


def deform_mesh(
    mesh: MeshData,
    amplitudes: tuple[float, float, float, float],
    morph_seed: tuple[int, int, int, int],
    deform_blend: float,
    noise_amount: float,
) -> MeshData:
    """Apply hash-driven deformation to a mesh.

    Uses 4 spherical-harmonic-like displacement modes along normals,
    plus a high-frequency noise displacement.

    Args:
        mesh: Base mesh to deform.
        amplitudes: Weights for 4 deformation modes (0-0.3 each).
        morph_seed: 4 seed bytes for noise generation.
        deform_blend: 0=no deformation, 1=full deformation (animated).
        noise_amount: High-frequency noise displacement magnitude (animated).
    """
    verts = mesh.vertices.copy()
    normals = mesh.normals.copy()

    # Convert to spherical coords for spherical-harmonic-like modes
    x, y, z = verts[:, 0], verts[:, 1], verts[:, 2]
    r = np.sqrt(x**2 + y**2 + z**2)
    r = np.maximum(r, 1e-10)
    theta = np.arccos(np.clip(z / r, -1, 1))  # polar angle
    phi = np.arctan2(y, x)  # azimuthal angle

    # 4 deformation modes (simplified spherical harmonics)
    mode0 = np.cos(2 * theta)  # Y_2^0 like
    mode1 = np.sin(theta) * np.cos(phi)  # Y_1^1 like
    mode2 = np.sin(2 * theta) * np.cos(2 * phi)  # Y_2^2 like
    mode3 = np.cos(3 * theta)  # Y_3^0 like

    displacement = (
        amplitudes[0] * mode0
        + amplitudes[1] * mode1
        + amplitudes[2] * mode2
        + amplitudes[3] * mode3
    )

    # Add pseudo-random noise using morph_seed
    seed_val = sum(s << (i * 8) for i, s in enumerate(morph_seed))
    rng = np.random.RandomState(seed_val)
    noise = rng.randn(len(verts)) * noise_amount

    # Apply displacement along normals, scaled by deform_blend
    total_disp = (displacement * deform_blend + noise)[:, np.newaxis]
    deformed = verts + normals * total_disp

    # Recompute normals
    new_normals = _compute_normals(deformed, mesh.indices)

    return MeshData(vertices=deformed.astype(np.float32), normals=new_normals, indices=mesh.indices)


def generate_mesh(shape: str, subdivision: int = 3) -> MeshData:
    """Generate a mesh by shape name."""
    match shape:
        case "icosphere":
            return icosphere(subdivision)
        case "torus":
            return torus()
        case "superellipsoid":
            return superellipsoid(e1=0.6, e2=0.6)
        case "octahedron":
            return superellipsoid(e1=1.5, e2=1.5)
        case "twisted_torus":
            return torus(major_radius=1.0, minor_radius=0.3,
                         major_segments=64, minor_segments=16)
        case _:
            return icosphere(subdivision)

File: service/src/geometry.rs
/// Procedural 3D mesh generation and deformation.
///
/// Port of Python `geometry.py`. Generates icosphere, torus, and
/// superellipsoid meshes with hash-driven vertex deformation.

use std::collections::HashMap;

use glam::Vec3;
use rand::SeedableRng;
use rand_chacha::ChaCha8Rng;
use rand_distr::{Distribution, StandardNormal};

use crate::params::Shape;

#[derive(Debug, Clone)]
pub struct MeshData {
    pub vertices: Vec<Vec3>,
    pub normals: Vec<Vec3>,
    pub indices: Vec<[u32; 3]>,
}

pub fn generate_mesh(shape: Shape, subdivision: u32) -> MeshData {
    match shape {
        Shape::Icosphere => icosphere(subdivision),
        Shape::Torus => torus(1.0, 0.4, 48, 24),
        Shape::Superellipsoid => superellipsoid(0.6, 0.6, 32, 32),
        Shape::Octahedron => superellipsoid(1.5, 1.5, 32, 32),
        Shape::TwistedTorus => torus(1.0, 0.3, 64, 16),
    }
}

pub fn icosphere(subdivision: u32) -> MeshData {
    let t = (1.0 + 5.0_f32.sqrt()) / 2.0;

    let raw_verts: [(f32, f32, f32); 12] = [
        (-1.0, t, 0.0), (1.0, t, 0.0), (-1.0, -t, 0.0), (1.0, -t, 0.0),
        (0.0, -1.0, t), (0.0, 1.0, t), (0.0, -1.0, -t), (0.0, 1.0, -t),
        (t, 0.0, -1.0), (t, 0.0, 1.0), (-t, 0.0, -1.0), (-t, 0.0, 1.0),
    ];

    let mut vertices: Vec<Vec3> = raw_verts
        .iter()
        .map(|&(x, y, z)| Vec3::new(x, y, z).normalize())
        .collect();

    let mut triangles: Vec<[u32; 3]> = vec![
        [0,11,5],[0,5,1],[0,1,7],[0,7,10],[0,10,11],
        [1,5,9],[5,11,4],[11,10,2],[10,7,6],[7,1,8],
        [3,9,4],[3,4,2],[3,2,6],[3,6,8],[3,8,9],
        [4,9,5],[2,4,11],[6,2,10],[8,6,7],[9,8,1],
    ];

    for _ in 0..subdivision {
        let mut midpoint_cache: HashMap<(u32, u32), u32> = HashMap::new();
        let mut new_triangles = Vec::with_capacity(triangles.len() * 4);

        for tri in &triangles {
            let a = get_midpoint(&mut vertices, &mut midpoint_cache, tri[0], tri[1]);
            let b = get_midpoint(&mut vertices, &mut midpoint_cache, tri[1], tri[2]);
            let c = get_midpoint(&mut vertices, &mut midpoint_cache, tri[2], tri[0]);
            new_triangles.push([tri[0], a, c]);
            new_triangles.push([tri[1], b, a]);
            new_triangles.push([tri[2], c, b]);
            new_triangles.push([a, b, c]);
        }
        triangles = new_triangles;
    }

    let normals = vertices.clone(); // For unit sphere, normal == position
    MeshData { vertices, normals, indices: triangles }
}

fn get_midpoint(
    vertices: &mut Vec<Vec3>,
    cache: &mut HashMap<(u32, u32), u32>,
    i1: u32,
    i2: u32,
) -> u32 {
    let key = (i1.min(i2), i1.max(i2));
    if let Some(&idx) = cache.get(&key) {
        return idx;
    }
    let mid = ((vertices[i1 as usize] + vertices[i2 as usize]) / 2.0).normalize();
    let idx = vertices.len() as u32;
    vertices.push(mid);
    cache.insert(key, idx);
    idx
}

pub fn torus(major_radius: f32, minor_radius: f32, major_seg: u32, minor_seg: u32) -> MeshData {
    let mut vertices = Vec::new();
    let mut indices = Vec::new();

    for i in 0..major_seg {
        let theta = 2.0 * std::f32::consts::PI * i as f32 / major_seg as f32;
        let (cos_t, sin_t) = (theta.cos(), theta.sin());
        for j in 0..minor_seg {
            let phi = 2.0 * std::f32::consts::PI * j as f32 / minor_seg as f32;
            let (cos_p, sin_p) = (phi.cos(), phi.sin());
            let r = major_radius + minor_radius * cos_p;
            vertices.push(Vec3::new(r * cos_t, r * sin_t, minor_radius * sin_p));
        }
    }

    for i in 0..major_seg {
        for j in 0..minor_seg {
            let curr = i * minor_seg + j;
            let next_j = i * minor_seg + (j + 1) % minor_seg;
            let next_i = ((i + 1) % major_seg) * minor_seg + j;
            let next_ij = ((i + 1) % major_seg) * minor_seg + (j + 1) % minor_seg;
            indices.push([curr, next_i, next_ij]);
            indices.push([curr, next_ij, next_j]);
        }
    }

    let normals = compute_normals(&vertices, &indices);
    MeshData { vertices, normals, indices }
}

pub fn superellipsoid(e1: f32, e2: f32, u_seg: u32, v_seg: u32) -> MeshData {
    fn sign_pow(base: f32, exp: f32) -> f32 {
        if base == 0.0 { return 0.0; }
        base.signum() * base.abs().powf(exp)
    }

    let mut vertices = Vec::new();
    for i in 0..=u_seg {
        let u = -std::f32::consts::FRAC_PI_2
            + std::f32::consts::PI * i as f32 / u_seg as f32;
        let (cos_u, sin_u) = (u.cos(), u.sin());
        for j in 0..=v_seg {
            let v = -std::f32::consts::PI
                + 2.0 * std::f32::consts::PI * j as f32 / v_seg as f32;
            let (cos_v, sin_v) = (v.cos(), v.sin());
            let x = sign_pow(cos_u, e1) * sign_pow(cos_v, e2);
            let y = sign_pow(cos_u, e1) * sign_pow(sin_v, e2);
            let z = sign_pow(sin_u, e1);
            vertices.push(Vec3::new(x, y, z));
        }
    }

    let mut indices = Vec::new();
    let row = v_seg + 1;
    for i in 0..u_seg {
        for j in 0..v_seg {
            let curr = i * row + j;
            indices.push([curr, curr + row, curr + row + 1]);
            indices.push([curr, curr + row + 1, curr + 1]);
        }
    }

    let normals = compute_normals(&vertices, &indices);
    MeshData { vertices, normals, indices }
}

fn compute_normals(vertices: &[Vec3], indices: &[[u32; 3]]) -> Vec<Vec3> {
    let mut normals = vec![Vec3::ZERO; vertices.len()];
    for tri in indices {
        let v0 = vertices[tri[0] as usize];
        let v1 = vertices[tri[1] as usize];
        let v2 = vertices[tri[2] as usize];
        let face_normal = (v1 - v0).cross(v2 - v0);
        for &idx in tri {
            normals[idx as usize] += face_normal;
        }
    }
    for n in &mut normals {
        let len = n.length();
        if len > 1e-10 {
            *n /= len;
        }
    }
    normals
}

pub fn deform_mesh(
    mesh: &MeshData,
    amplitudes: &[f32; 4],
    morph_seed: &[u8; 4],
    deform_blend: f32,
    noise_amount: f32,
) -> MeshData {
    let mut verts = mesh.vertices.clone();

    // Spherical-harmonic-like modes
    for (i, vert) in verts.iter_mut().enumerate() {
        let v = mesh.vertices[i];
        let n = mesh.normals[i];
        let r = v.length().max(1e-10);
        let theta = (v.z / r).clamp(-1.0, 1.0).acos();
        let phi = v.y.atan2(v.x);

        let mode0 = (2.0 * theta).cos();
        let mode1 = theta.sin() * phi.cos();
        let mode2 = (2.0 * theta).sin() * (2.0 * phi).cos();
        let mode3 = (3.0 * theta).cos();

        let displacement = amplitudes[0] * mode0
            + amplitudes[1] * mode1
            + amplitudes[2] * mode2
            + amplitudes[3] * mode3;

        *vert = v + n * displacement * deform_blend;
    }

    // Add noise
    if noise_amount > 0.0 {
        let mut seed = [0u8; 32];
        for i in 0..32 {
            seed[i] = morph_seed[i % 4].wrapping_add(i as u8);
        }
        let mut rng = ChaCha8Rng::from_seed(seed);
        for (i, vert) in verts.iter_mut().enumerate() {
            let noise: f32 = StandardNormal.sample(&mut rng);
            *vert += mesh.normals[i] * noise * noise_amount;
        }
    }

    let normals = compute_normals(&verts, &mesh.indices);
    MeshData {
        vertices: verts,
        normals,
        indices: mesh.indices.clone(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_icosphere_vertex_count() {
        for n in 0..4 {
            let mesh = icosphere(n);
            let expected = 10 * 4u32.pow(n) + 2;
            assert_eq!(mesh.vertices.len() as u32, expected,
                       "subdivision {n}: got {}, expected {expected}", mesh.vertices.len());
        }
    }

    #[test]
    fn test_icosphere_unit_sphere() {
        let mesh = icosphere(2);
        for v in &mesh.vertices {
            assert!((v.length() - 1.0).abs() < 1e-5);
        }
    }

    #[test]
    fn test_torus_nonempty() {
        let mesh = torus(1.0, 0.4, 48, 24);
        assert!(!mesh.vertices.is_empty());
        assert!(!mesh.indices.is_empty());
    }

    #[test]
    fn test_deform_zero_blend() {
        let mesh = icosphere(2);
        let deformed = deform_mesh(&mesh, &[0.1; 4], &[0; 4], 0.0, 0.0);
        for (orig, def) in mesh.vertices.iter().zip(deformed.vertices.iter()) {
            assert!((orig - def).length() < 1e-6);
        }
    }
}
